tm2/pkg/sdk/auth/keeper.go

```diff
@@ -160,3 +164,148 @@ func (ak AccountKeeper) decodeAccount(bz []byte) (acc std.Account) {
 	}
 	return
 }
+
+type GasPriceContextKey struct{}
+
+type GasPriceKeeper struct {
+	key store.StoreKey
+}
+
+// GasPriceKeeper
+// The GasPriceKeeper stores the history of gas prices and calculates
+// new gas price with formula parameters
+func NewGasPriceKeeper(key store.StoreKey) GasPriceKeeper {
+	return GasPriceKeeper{
+		key: key,
+	}
+}
+
+// SetGasPrice is called in InitChainer to store initial gas price set in the genesis
+
+func (gk GasPriceKeeper) SetGasPrice(ctx sdk.Context, gp std.GasPrice) {
+	if (gp == std.GasPrice{}) {
+		return
+	}
+	stor := ctx.Store(gk.key)
+	bz, err := amino.Marshal(gp)
+	if err != nil {
+		panic(err)
+	}
+	stor.Set([]byte(GasPriceKey), bz)
+}
+
+// We store the history. If the formula changes, we can replay blocks
+// and apply the formula to a specific block range. The new gas price is
+// calculated in EndBlock().
+func (gk GasPriceKeeper) UpdateGasPrice(ctx sdk.Context) {
+	params := ctx.Value(AuthParamsContextKey{}).(Params)
+	gasUsed := ctx.BlockGasMeter().GasConsumed()
+	maxBlockGas := ctx.ConsensusParams().Block.MaxGas
+	lgp := gk.LastGasPrice(ctx)
+	newGasPrice := gk.calcBlockGasPrice(lgp, gasUsed, maxBlockGas, params)
+	gk.SetGasPrice(ctx, newGasPrice)
+}
+
+// calcBlockGasPrice calculates the minGasPrice for the txs to be included in the next block.
+// newGasPrice = lastPrice + lastPrice*(gasUsed-TargetBlockGas)/TargetBlockGas/GasCompressor)
+//
+// The math formula is an abstraction of a simple solution for the underlying problem we're trying to solve.
+// 1. What do we do if the gas used is less than the target gas in a block?
+// 2. How do we bring the gas used back to the target level, if gas used is more than the target?
+// We simplify the solution with a one-line formula to explain the idea. However, in reality, we need to treat
+// two scenarios differently. For example, in the first case, we need to increase the gas by at least 1 unit,
+// instead of round down for the integer divisions, and in the second case, we should set a floor
+// as the target gas price. This is just a starting point. Down the line, the solution might not be even
+// representable by one simple formula
+
+func (gk GasPriceKeeper) calcBlockGasPrice(lastGasPrice std.GasPrice, gasUsed int64, maxGas int64, params Params) std.GasPrice {
+	// If no block gas price is set, there is no need to change the last gas price.
+	if lastGasPrice.Price.Amount == 0 {
+		return lastGasPrice
+	}
+
+	// This is also a configuration to indicate that there is no need to change the last gas price.
+	if params.TargetGasRatio == 0 {
+		return lastGasPrice
+	}
+	// if no gas used, no need to change the lastPrice
+	if gasUsed == 0 {
+		return lastGasPrice
+	}
+	var (
+		num   = new(big.Int)
+		denom = new(big.Int)
+	)
+
+	// targetGas = maxGax*TargetGasRatio/100
+
+	num.Mul(big.NewInt(maxGas), big.NewInt(params.TargetGasRatio))
```

Comment:
Have you thought about using hooliman's uint256, instead of `big.Int`?

Reply:
[holiman's uint256](https://github.com/holiman/uint256)  is awesome. However, I would take a conservative approach when using low-level libraries like this. Plus, the calculation here only happens on a per-block basis."

Reply:
[Holiman's uint256](https://github.com/holiman/uint256) is awesome. It is a fixed-sized uint and perfect for Ethereum RLP and the EVM. However, it can overflow and underflow, and I don't think it is used for ETH gas calculations either.

With big.Int, we don't need to worry about overflow or underflow during gas calculations since its size is dynamic. Plus, we only calculate this number on a per-block basis, so the trade-off is minimal for the peace of mind it provides.